openapi-model/src/main/scala/sttp/apispec/openapi/validation/OpenAPIComparator.scala

```diff
@@ -0,0 +1,277 @@
+    if (issues.isEmpty)
+      None
+    else
+      Some(IncompatibleOperation(httpMethod, issues))
+  }
+
+  private def checkParameter(clientParameter: Parameter, serverParameter: Parameter): Option[IncompatibleParameter] = {
+    val isCompatibleStyle = serverParameter.style == clientParameter.style
+    val isCompatibleExplode = serverParameter.explode == clientParameter.explode
+    val isCompatibleAllowEmptyValue = serverParameter.allowEmptyValue == clientParameter.allowEmptyValue
+    val isCompatibleAllowReserved = serverParameter.allowReserved == clientParameter.allowReserved
+
+    val issues =
+      checkSchema(clientParameter.schema, serverParameter.schema).toList ++
+        checkContent(clientParameter.content, serverParameter.content).toList ++
+        (if (!isCompatibleStyle) Some(IncompatibleStyle()) else None).toList ++
+        (if (!isCompatibleExplode) Some(IncompatibleExplode()) else None).toList ++
+        (if (!isCompatibleAllowEmptyValue) Some(IncompatibleAllowEmptyValue()) else None).toList ++
+        (if (!isCompatibleAllowReserved) Some(IncompatibleAllowReserved()) else None).toList
+
+    if (issues.isEmpty)
+      None
+    else
+      Some(IncompatibleParameter(clientParameter.name, issues))
+  }
+
+  private def checkContent(
+      clientContent: ListMap[String, MediaType],
+      serverContent: ListMap[String, MediaType]
+  ): Option[IncompatibleContent] = {
+    val issues = clientContent.flatMap { case (clientMediaType, clientMediaTypeDescription) =>
+      val serverMediaTypeDescription = serverContent.get(clientMediaType)
+      serverMediaTypeDescription match {
+        case None => Some(MissingMediaType(clientMediaType))
+        case Some(serverMediaTypeDescription) =>
+          checkMediaType(clientMediaType, clientMediaTypeDescription, serverMediaTypeDescription)
+      }
+    }
+
+    if (issues.isEmpty)
+      None
+    else
+      Some(IncompatibleContent(issues.toList))
+  }
+
+  private def checkMediaType(
+      mediaType: String,
+      clientMediaTypeDescription: MediaType,
+      serverMediaTypeDescription: MediaType
+  ): Option[IncompatibleMediaType] = {
+    val issues = checkSchema(clientMediaTypeDescription.schema, serverMediaTypeDescription.schema)
+    if (issues.nonEmpty)
+      Some(IncompatibleMediaType(mediaType, issues.toList))
+    else
+      None
+    // TODO: encoding?
+  }
+
+  private def checkSchema(
+      clientSchema: Option[SchemaLike],
+      serverSchema: Option[SchemaLike]
+  ): Option[OpenAPICompatibilityIssue] = {
+    (serverSchema, clientSchema) match {
+      case (Some(serverSchema), Some(clientSchema)) =>
+        val schemaComparator = new SchemaComparator(clientSchemas, serverSchemas)
+        val schemaIssues = schemaComparator.compare(clientSchema, serverSchema)
+        if (schemaIssues.nonEmpty)
+          Some(IncompatibleSchema(schemaIssues))
+        else
+          None
+      case _ => None
```

Comment:
isn't it an incompatibility if one schema is provided, and the other is not?

Reply:
I guess just `MissingSchema` is enough

Reply:
OpenAPI requires that a `Parameter` or `Header` object defines either a `schema` or `content` with exactly one entry. I guess it could technically be possible for two parameters/headers to be compatible when one of them has a `schema` while the other one has a `content`?